common/rpc/factory.go

```diff
@@ -37,13 +37,19 @@ import (
 	"github.com/uber/cadence/common/log"
 	"github.com/uber/cadence/common/log/tag"
 	"github.com/uber/cadence/common/membership"
+	"github.com/uber/cadence/common/service"
 )
 
 const (
 	defaultGRPCSizeLimit = 4 * 1024 * 1024
 	factoryComponentName = "rpc-factory"
 )
 
+var (
+	// P2P outbounds are only needed for history and matching services
+	servicesToTalkP2P = []string{service.History, service.Matching}
```

Comment:
Is frontend --> History not in scope? I can see a strong case for having frontend retain connections to history and matching instances as peers?

Reply:
I also think that worker might need need membership updates for the domain config stack (@Shaddoll can correct me)

Reply:
These are sharded services that all other services communicate via p2p connections. RPC factory subscribes to membership updates of these services for that reason. Frontend and worker will still use this rpc factory and will be able to communicate with matching&history via p2p